File: day1/java/balanced.java

import java.util.Scanner;
import java.util.Stack;

class Balanced{
    String exp;
    Balanced(String exp){
        this.exp = exp;
    }
    void chechbalance(){
        Stack<Character> stack = new Stack<Character>();
        for(int i=0; i<exp.length(); i++){
            char ch = exp.charAt(i);
            if(ch=='(' || ch=='{' || ch=='['){
                stack.push(ch);
            } else if(ch==')' || ch=='}' || ch==']'){
                if(stack.isEmpty()){
                    System.out.println("Not Balanced");
                    return;
                }
                char top = stack.pop();
                if((top=='(' && ch!=')') || (top=='{' && ch!='}') || (top=='[' && ch!=']')){
                    System.out.println("Not Balanced");
                    return;
                }
            }
        }
        if(stack.isEmpty()){
            System.out.println("Balanced");
        } else {
            System.out.println("Not Balanced");
        }
    }

}
class BalanceCheck{
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int c;
        String exp;
        do { 
            System.err.print("\nCoice 1: Check Balanced Parenthesis\nChoice 2: Exit\nEnter your choice: ");
            c = sc.nextInt();
            switch(c){
                case 1:
                    System.out.print("Enter the expression: ");
                    exp = sc.next();
                    Balanced b = new Balanced(exp);
                    b.chechbalance();
                    break;
                case 2:
                    System.out.println("Exiting...");
                    break;
                default:
                    System.out.println("Invalid Choice");
            }
        } while (c!=2);
    }
}

File: day1/Python/balanced.py
def is_balanced(expression):
    stack = []
    opening = set('([{')
    closing = set(')]}')
    matches = {')': '(', ']': '[', '}': '{'}

    for char in expression:
        if char in opening:
            stack.append(char)
        elif char in closing:
            if not stack or stack[-1] != matches[char]:
                return False
            stack.pop()

    return not stack

if __name__ == '__main__':
    n=1
    while n:
        n = input("Choice 1: Expression checking\nChoice 2: Exit\nEnter your choice: ")
        if n == '1':
            n = input("Enter the expression: ")
            print("Balanced" if(is_balanced(n)) else "Not balanced")
        elif n == '2':
            print("Exiting...")
            break
        else:
            print("Invalid choice. Please try again.")
